File: 07_Trees/11_Validate_Binary_Search_Tree.java
class Solution {

    public boolean dfs(TreeNode root, long minVal, long maxVal) {
        if (root == null)
            return true;
        if (root.val <= minVal || root.val >= maxVal)
            return false;
        return dfs(root.left, minVal, root.val) && dfs(root.right, root.val, maxVal);
    }

    public boolean isValidBST(TreeNode root) {
        long minVal = Integer.MIN_VALUE, maxVal = Integer.MAX_VALUE;
        minVal--;
        maxVal++;
        return dfs(root, minVal, maxVal);
    }
}

File: 07_Trees/11_Validate_Binary_Search_Tree.cpp
class Solution
{
public:
    bool dfs(TreeNode *root, long minVal, long maxVal)
    {
        if (!root)
            return true;
        if (root->val <= minVal || root->val >= maxVal)
            return false;
        return dfs(root->left, minVal, root->val) && dfs(root->right, root->val, maxVal);
    }

    bool isValidBST(TreeNode *root)
    {
        long minVal = INT_MIN, maxVal = INT_MAX;
        minVal--;
        maxVal++;

        return dfs(root, minVal, maxVal);
    }
};